include/internal/hashtable.h

```diff
@@ -139,6 +139,22 @@ memset((key), 0, sizeof(*(key))); \
    ossl_ht_strcase((key)->keyfields.member, value, sizeof((key)->keyfields.member) -1); \
 } while(0)
 
+/*
+ * Same as HT_SET_KEY_STRING but also takes length of the string.
+ */
+#define HT_SET_KEY_STRING_N(key, member, value, len) do { \
+    if ((value) != NULL) \
+        strncpy((key)->keyfields.member, value, len); \
+} while(0)
+
+/*
+ * This is the same as HT_SET_KEY_STRING_CASE but also takes length of the
+ * string.
+ */
+#define HT_SET_KEY_STRING_CASE_N(key, member, value, len) do { \
+   ossl_ht_strcase((key)->keyfields.member, value, len); \
+} while(0)
```

Comment:
This is a bit fragile I think.  When the hashtable defines a key with a string, we use HT_DEF_KEY_FIELD_CHAR_ARRAY() which accepts a size parameter to define a member element as:
```
unsigned char <name>[size]
```

Then using HT_SET_KEY_STRING copies the input string into the key->member.<name> array, limiting its size implicitly by the size of the declared array.  If we create HT_SET_KEY_STRING_CASE_N, we allow for the possibility that the input name that is copied will overrun that declared array.

Reply:
Solution would either to:
1. Only copy `min(len, sizeof((key)->keyfields.member -1)` bytes from name: Could truncate part of the name, but this would just be same as previous functionality anyways
2. Only do the copy if `len < sizeof((key)->keyfields.member)` and do nothing otherwise: Could be an unexpected no-op

So I'm thinking of just going with (1)